src/client/pythonEnvironments/discovery/locators/services/pipEnvHelper.ts

```diff
@@ -0,0 +1,151 @@
+    }
+    return undefined;
+}
+
+/**
+ * If interpreter path belongs to a pipenv environment which is located inside a project, return associated Pipfile,
+ * otherwise return `undefined`.
+ * @param interpreterPath Absolute path to any python interpreter.
+ */
+async function getPipfileIfLocal(interpreterPath: string): Promise<string | undefined> {
+    // Local pipenv environments are created by setting PIPENV_VENV_IN_PROJECT to 1, which always names the environment
+    // folder '.venv': https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_VENV_IN_PROJECT
+    // This is the layout we wish to verify.
+    // project
+    // |__ Pipfile  <--- check if Pipfile exists here
+    // |__ .venv    <--- check if name of the folder is '.venv'
+    //     |__ Scripts/bin
+    //         |__ python  <--- interpreterPath
+    const venvFolder = path.dirname(path.dirname(interpreterPath));
+    if (path.basename(venvFolder) !== '.venv') {
+        return undefined;
+    }
+    const directoryWhereVenvResides = path.dirname(venvFolder);
+    return _getAssociatedPipfile(directoryWhereVenvResides, { lookIntoParentDirectories: false });
+}
+
+/**
+ * Returns the project directory for pipenv environments given the environment folder
+ * @param envFolder Path to the environment folder
+ */
+async function getProjectDir(envFolder: string): Promise<string | undefined> {
+    // Global pipenv environments have a .project file with the absolute path to the project
+    // See https://github.com/pypa/pipenv/blob/v2018.6.25/CHANGELOG.rst#features--improvements
+    // This is the layout we expect
+    // <Environment folder>
+    // |__ .project  <--- check if .project exists here
+    // |__ Scripts/bin
+    //     |__ python  <--- interpreterPath
+    // We get the project by reading the .project file
+    const dotProjectFile = path.join(envFolder, '.project');
+    if (!(await pathExists(dotProjectFile))) {
+        return undefined;
+    }
+    const projectDir = await readFile(dotProjectFile);
+    if (!(await pathExists(projectDir))) {
+        return undefined;
+    }
+    return projectDir;
+}
+
+/**
+ * If interpreter path belongs to a global pipenv environment, return associated Pipfile, otherwise return `undefined`.
+ * @param interpreterPath Absolute path to any python interpreter.
+ */
+async function getPipfileIfGlobal(interpreterPath: string): Promise<string | undefined> {
+    const envFolder = path.dirname(path.dirname(interpreterPath));
+    const project = await getProjectDir(envFolder);
+    if (project === undefined) {
+        return undefined;
+    }
+
+    // This is the layout we expect to see.
+    //  project
+    // |__ Pipfile  <--- check if Pipfile exists here and return it
+    // The name of the project (directory where Pipfile resides) is used as a prefix in the environment folder
+    const envFolderName = path.basename(envFolder);
+    if (!envFolderName.startsWith(`${path.basename(project)}-`)) {

```

Comment:
This is probably fine.  However, users do weird things sometimes and it often helps to play it safe.  A little extra effort now can save a few users headaches down the road and save us from having to triage a few extra issues.  With a regular expression we can avoid matching things we shouldn't.

```suggestion
    if (envFolderName.match(new RegExp(`^${path.basename(project)}-[0-9a-fA-F]+$`)) === null) {
```

Reply:
Alright, I have changed it.

Reply:
Something seems to be incorrect, some tests are failing. I am not very aware of the regex syntax

Reply:
Something is wrong in the pattern suggested.

`const matches = envFolderName.match(new RegExp(`^${path.basename(projectDir)}-[0-9a-fA-F]+$`))`

`matches` is `null` here when `projectDir = path\to\project` and `envFolderName = project-2s1eXEJ2`.

I don't want to spend too much time here so I'm reverting to the original way.

Reply:
`^project-[0-9a-fA-F]+$` doesn't match with `project-2s1eXEJ2` but removing `$` at the end of the regex possibly corrects the regex. I don't quite understand.

Reply:
The pattern I used assumed the hash was hexadecimal.  If it is instead any arbitrary non-whitespace character than the pattern would end with `-\S+$`.

Reply:
It is always hexadecimal.

> matches is null here when projectDir = path\to\project and envFolderName = project-2s1eXEJ2.

Since we're not able to easily guess the pattern in that case, I'm going with the old way.

Reply:
It is always hexadecimal. (For eg. `project-2s1eXEJ2`)

> matches is null here when projectDir = path\to\project and envFolderName = project-2s1eXEJ2.

Since we're not able to easily guess the correct pattern for hexadecimal in this case, I'm going with the old way.